turbopack/crates/turbo-tasks-backend/src/interning_serde/mod.rs

```diff
@@ -0,0 +1,40 @@
+//! Exposed for usage in `turbo-tasks-backend`
+
+use std::io::Write;
+
+use indexmap::IndexSet;
+use rustc_hash::FxBuildHasher;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use turbo_rcstr::RcStr;
+
+#[derive(Serialize, Deserialize)]
+struct Data(Vec<u8>, IndexSet<RcStr, FxBuildHasher>);
+
+pub fn to_vec<T>(config: &pot::Config, value: &T) -> pot::Result<Vec<u8>>
+where
+    T: Serialize,
+{
+    let mut vec = Vec::new();
+    to_writer(config, value, &mut vec)?;
+    Ok(vec)
+}
+
+pub fn to_writer<T, W>(config: &pot::Config, value: &T, writer: W) -> pot::Result<()>
+where
+    T: Serialize,
+    W: Write,
+{
+    let (result, ser_map) = turbo_rcstr::set_ser_map(|| config.serialize(value));
+    let value = result?;
+    let data = Data(value, ser_map);
+    config.serialize_into(&data, writer)
+}
+
+pub fn from_slice<T>(config: &pot::Config, slice: &[u8]) -> pot::Result<T>
+where
+    T: DeserializeOwned,
+{
+    let data: Data = config.deserialize(slice)?;
```

Comment:
This works, but it could be a bit more efficient if it would borrow `Data.0` from the slice. It also doesn't need to to hash all the deserialized RcStr due to the IndexSet.

I would recommend to split `Data` into `SerData` and `DeserData`. And I would use some SmallVec to make (de)serializing small values allocate only on the stack.

```
#[derive(Serialize)]
struct SerData(SmallVec<[u8; 16]>, FxIndexSet<RcStr>);

#[derive(Deserialize)]
struct DeserData<'l>(&'l [u8; 16]>, SmallVec<[RcStr; 16]>);
```